File: BlogModule.Application/QueryHandlers/Tag/GetTags/GetTagsQueryHandler.cs
using BlogModule.Application.Contract.Tags.GetTags;
using BlogModule.Domain.RepositoryInterfaces.Tag;
using Results.Domain;
using Shared.CQRS;
using Shared.CQRS.QueryHandlers;

namespace BlogModule.Application.QueryHandlers.Tag.GetTags;

public class GetTagsQueryHandler : IQueryHandler<GetTagsQuery, HandlerResult<GetTagsResponse, IErrorResult>>
{
    private readonly ITagQueryRepository _repository;
        
    public GetTagsQueryHandler(ITagQueryRepository repository)
    {
        _repository = repository;
    }
        
    public async Task<HandlerResult<GetTagsResponse, IErrorResult>> Handle(GetTagsQuery request, CancellationToken cancellationToken)
    {
        var tags = await _repository.GetAllDistinctAsync(
            request.OrderBy, 
            request.OrderDirection,
            request.BlogPostId,
            cancellationToken);

        var response = new GetTagsResponse() {
            OrderBy = request.OrderBy,
            OrderDirection = request.OrderDirection,
            Items = tags
        };
            
        return response;
    }
}

File: BlogApi/Application/query_handlers/blog_post_tag/get_tags/get_tags_query_handler.go
package queries

import (
	contract "blog-api/Application.contract/Tags/GetTags"
	repository_interfaces "blog-api/Domain/repository_interfaces/blog_post_tag_repository"
	"blog-api/Domain/responses"
	"context"
)

type GetTagsQueryHandler struct {
	blogPostTagQueryRepository repository_interfaces.BlogPostTagQueryRepository
}

func NewGetTagsQueryHandler(
	blogPostTagQueryRepository repository_interfaces.BlogPostTagQueryRepository,
) *GetTagsQueryHandler {
	return &GetTagsQueryHandler{
		blogPostTagQueryRepository: blogPostTagQueryRepository,
	}
}

func (h *GetTagsQueryHandler) Handle(ctx context.Context, query *GetTagsQuery) (*contract.GetTagsResponse, error) {
	tags, err := h.blogPostTagQueryRepository.GetTagsByBlogPostID(query.BlogPostId, query.Sortable)
	if err != nil {
		response := responses.NewResponse[contract.GetTagsResponse](500, "Failed to retrieve tags")
		return &response, nil
	}

	return &contract.GetTagsResponse{
		BaseResponse: responses.NewBaseResponse(200, "Tags retrieved successfully"),
		Items:        *tags,
	}, nil
}
